### PR TITLE
Support activation sharding

### PR DESCRIPTION
**Pull request by wonjoo-wj**
With https://github.com/pytorch/xla/pull/6524, we can now support activation sharding. 

This is an example of the changes we need to call the new dynamo python custom op. 

Note that this requires https://github.com/pytorch/xla/pull/6524 to be merged first. Tested locally, llama 2 inference with dynamo+spmd (with activation sharding) is successful: https://gist.github.com/wonjoolee95/a290a68f29c52bd395b16ae6df651531.

Also note that performance is not tested, so we need to find the optimal sharding strategy. This only tests the functionality.

### CODE REVIEW

File: llama/model.py [hunk 1]
```diff
@@ -205,10 +205,20 @@ def forward(
 
         # # Activation output sharding
         # # TODO(yeounoh) remove this after activation sharding support is enabled.
-        num_devices = xr.global_runtime_device_count()
-        device_ids = torch.arange(num_devices)
-        data_model_mesh = xs.Mesh(device_ids, (4, 1, 2))
-        xs.mark_sharding(output, data_model_mesh, (0, 1, 2), use_dynamo_custom_op=True)
+        # num_devices = xr.global_runtime_device_count()
```
Comment (miladm) [inline llama/model.py:208]: @wonjoolee95  can you add a discussion on the reason to keep `output = self.wo(output)` in line 204 when we want to do the `dynamo_mark_sharding` call  after?

Comment (yeounoh) [reply llama/model.py:208]: This is the original implementation, and we are just adding the annotations. We have to do output projection as-is.

File: llama/model.py [hunk 2]
```diff
@@ -205,10 +205,20 @@ def forward(
...
+        # device_ids = torch.arange(num_devices)
+        # data_model_mesh = xs.Mesh(device_ids, (4, 1, 2))
+        # xs.mark_sharding(output, data_model_mesh, (0, 1, 2), use_dynamo_custom_op=True)
+
+        # custom python dynamo mark sharding
+        import torch_xla.experimental.dynamo_mark_sharding
+        device_ids = [0]
+        mesh_shape = [1, 1, 1]
```
Comment (yeounoh) [inline llama/model.py:216]: wait, why is the mesh shape 1, 1, 1?

**Review by yeounoh — CHANGES_REQUESTED** [review]
This is for testing -- @wonjoolee95 
- could you remove the "[WONJOO]" debugging probes, and make it look more formal if useful? 
- Also, you have to change the mesh shape. Don't think we are sharding the activation in the current impl?

**Comment by wonjoo-wj** [comment]
Moving this over to https://github.com/pytorch-tpu/llama/pull/55/

**Pull request closed without merging**